### ISSUE TITLE
Reconsidering the `no-autofocus` rule

### ISSUE THREAD

**Issue by smhigley** [jsx-eslint/eslint-plugin-jsx-a11y#1093]
I know "don't use `autofocus`" has been a web a11y maxim for a long time, but with consideration I think it has been outdated for a while now, and mostly lasted through inertia. While we still definitely don't want devs to set `autofocus` to elements that render on page load, dynamically rendered popups, dialogs, and similar UI have become normal, and `autofocus` is a perfectly valid (and sometimes best practice) way to handle moving focus within them when they appear.

I believe the original "don't use autofocus" recommendation was more useful back in the days when the web was intended to be largely static, and valid usage was more of an outlier. I wonder if that's still useful nowadays, especially since I haven't encountered a mistaken `autofocus` in quite some time, and valid use cases around declaratively managing focus are quite common.

To that point, the relatively recent `<dialog>` element and `popover` attribute both explicitly recommend authors use `autofocus` to set where focus should land when they open. The recommended use of `autofocus` is also likely to increase, since a big part of Open UI's goal is to incubate web platform features that allow authors to create highly interactive UI declaratively in HTML, and `autofocus` as a declarative focus management attribute is quite useful.

For reference, here are places autofocus is mentioned in dialog/popover incubation info and on MDN:
- [Open UI explainer for popover](https://open-ui.org/components/popover.research.explainer/#focus-management)
- [HTML spec for popover](https://html.spec.whatwg.org/multipage/popover.html#the-popover-attribute)
- [HTML spec for dialog](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-dialog-element)
- [MDN page for dialog](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/dialog#accessibility)

The valid use also isn't limited to dialog & popover -- another place I used it recently was on a button that rendered an input when invoked -- the easiest and most robust way to set focus in that instance is with `autofocus`. Obviously it can still be abused, but I'm not sure it's any more dangerous or less valid than `tabindex` or `focusgroup` or `aria-hidden`.

Based on all the above, I think it's a recommendation that the web has now outgrown and should likely be removed. But please let me know if there's anything I'm missing or should reconsider.

**Comment by ljharb** [comment]
I don't agree. "don't use autofocus" means "don't use the html attribute". It's perfectly fine to manually `.focus()` on a text field in a popup or similar.

**Comment by smhigley** [comment]
I'm also talking about the HTML attribute, not `element.focus()` -- it's the explicit best practice for HTML dialogs and popovers.

Could you elaborate on why you think modern web devs shouldn't use it?

**Comment by ljharb** [comment]
Can you elaborate on where it’s an explicit best practice to use it?

**Comment by smhigley** [comment]
Sure! I'll pull the relevant quotes or examples from the explainers and specs linked above:

in the [Open UI explainer for popover](https://open-ui.org/components/popover.research.explainer/#focus-management):
> Additionally, if the top layer element should receive immediate focus, there is a question about which part of the element gets that initial focus. For example, the element itself could receive focus, or one of its focusable descendants could receive focus.
>
> To provide control over these behaviors, the autofocus attribute can be used on or within popovers. When present on a popover or one of its descendants, it will result in focus being moved to the popover or the specified element when the popover is shown. 

In the [HTML popover spec](https://html.spec.whatwg.org/multipage/popover.html#the-popover-attribute), the second example goes over how to handle focus:
> The following shows the baseline markup to create a custom menu popover, where the first menuitem will receive keyboard focus when the popover is invoked due to the use of the autofocus attribute. [...]
```html
<button popovertarget=m>Actions</button>
<div role=menu id=m popover>
  <button role=menuitem tabindex=-1 autofocus>Edit</button>
  <button role=menuitem tabindex=-1>Hide</button>
  <button role=menuitem tabindex=-1>Delete</button>
</div>
```

And further down, there is the [popover focusing steps](https://html.spec.whatwg.org/multipage/popover.html#popover-focusing-steps) section that outlines how `autofocus` is handled.

The [HTML dialog spec](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-dialog-element) has this as its third paragraph, followed by an example demonstrating `autofocus` usage:
> An important part of user-facing dialog behavior is the placement of initial focus. The [dialog focusing steps](https://html.spec.whatwg.org/multipage/interactive-elements.html#dialog-focusing-steps) attempt to pick a good candidate for initial focus when a dialog is shown, but might not be a substitute for authors carefully thinking through the correct choice to match user expectations for a specific dialog. As such, authors should use the [autofocus](https://html.spec.whatwg.org/multipage/interaction.html#attr-fe-autofocus) attribute on the descendant element of the dialog that the user is expected to immediately interact with after the dialog opens. If there is no such element, then authors should use the [autofocus](https://html.spec.whatwg.org/multipage/interaction.html#attr-fe-autofocus) attribute on the [dialog](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-dialog-element) element itself.

The MDN page for `<dialog>` begins its [accessibility section](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/dialog#accessibility) with this:
> When implementing a dialog, it is important to consider the most appropriate place to set user focus. When using [HTMLDialogElement.showModal()](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/showModal) to open a <dialog>, focus is set on the first nested focusable element. Explicitly indicating the initial focus placement by using the [autofocus](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Global_attributes/autofocus) attribute will help ensure initial focus is set on the element deemed the best initial focus placement for any particular dialog.

Outside of popover/dialog, using `autofocus` is both ergonomic because it doesn't require script, and especially in cases where the to-be-focused element renders a tick after it is triggered (e.g. as a result of a state change in React) it means you wouldn't need to either worry about timing or create an additional `useEffect` or similar.

`autofocus` is also the recommended focus management method for newer Open UI proposals such as the [`openable` API](https://open-ui.org/components/openable.explainer/) and likely [overscroll](https://open-ui.org/components/overscroll-actions.explainer/) as well, for cases when the overscrolled UI should take focus. Since a primary goal is to allow for authoring declarative richly interactive UI in HTML, the `autofocus` attribute is key to creating any dynamic surface that needs to take focus when shown.

Are there any specific concerns you have with using `autofocus` that you'd want raised with Open UI or HTML, or is it more with usage outside of the context of those elements & attributes?

**Comment by smhigley** [comment]
Just wanted to check in on this issue -- any updates or comments? For context, I'm a member of the w3c & Open UI, and contribute to a number of the cited specs, which is why I'm in the weeds of changes to `autofocus` usage and recommendations 😅. And of course use this eslint plugin too.

I'm also happy to contribute a PR for this as well, if there's agreement on it.
